lib/services/role.go

```diff
@@ -1165,23 +1169,31 @@ func (set RoleSet) PinSourceIP() bool {
 	return false
 }
 
-// GetAccessState returns the AccessState for the user given their roles, the
-// cluster auth preference, and whether MFA and the user's device were verified.
+// GetAccessState returns the AccessState, setting [AccessState.MFARequired]
+// according to the user's roles and cluster auth preference.
 func (set RoleSet) GetAccessState(authPref types.AuthPreference) AccessState {
+	return AccessState{
+		MFARequired: set.getMFARequired(authPref.GetRequireMFAType()),
+		// We don't set ClusterDeviceMode here, as both ClusterDeviceMode and
+		// DeviceVerified should be set in tandem.
+	}
+}
+
+func (set RoleSet) getMFARequired(clusterRequireMFAType types.RequireMFAType) MFARequired {
 	// per-session MFA is overridden by hardware key PIV touch requirement.
 	// check if the auth pref or any roles have this option.
-	if authPref.GetRequireMFAType() == types.RequireMFAType_HARDWARE_KEY_TOUCH {
-		return AccessState{MFARequired: MFARequiredNever}
+	if clusterRequireMFAType == types.RequireMFAType_HARDWARE_KEY_TOUCH {
+		return MFARequiredNever
 	}
 	for _, role := range set {
 		if role.GetOptions().RequireMFAType == types.RequireMFAType_HARDWARE_KEY_TOUCH {
-			return AccessState{MFARequired: MFARequiredNever}
+			return MFARequiredNever
```

Comment:
Yes; HARDWARE_KEY_TOUCH means you have to touch the Yubikey to get access to the private key within, so it effectively pushes the MFA check to client-side. We know the setting happens because of YK attestation, so it's alright.